ci/run_all_sets_parallel.php

```diff
@@ -0,0 +1,213 @@
+        if ($noMoreProcessesAreLeft || $maxNumberOfProcessesAreRunning) {
+            sleep($secondsToSleep);
+        }
+    }
+
+    private function canStartAnotherProcess(array $remainingTasks, array $runningProcesses, int $max): bool
+    {
+        $hasOpenTasks              = count($remainingTasks) > 0;
+        $moreProcessesCanBeStarted = count($runningProcesses) < $max;
+        return $hasOpenTasks && $moreProcessesCanBeStarted;
+    }
+
+    private function createProcess(Task $task): Process
+    {
+        $command = ['php', 'bin/rector', 'process', $task->getPathToFile(), '--set', $task->getSetName(), '--dry-run'];
+
+        return new Process($command, __DIR__.'/..');
+    }
+
+    private function printSuccess(string $set, int $totalTasks, int $finishedTasks): void
+    {
+        echo sprintf('✔ Set "%s" is OK (%2d / %2d)'.PHP_EOL, $set, $finishedTasks, $totalTasks);
+    }
+
+    private function printError(string $set, Throwable $t, int $totalTasks, int $finishedTasks): void
+    {
+        echo sprintf('❌ Set "%s" failed: %s (%2d / %2d)'.PHP_EOL, $set, $t->getMessage(), $finishedTasks, $totalTasks);
+    }
+
+    private function printInfo(array $tasks, int $maxProcesses): void
+    {
+        echo sprintf('Running %d sets with %d parallel processes'.PHP_EOL.PHP_EOL, count($tasks), $maxProcesses);
+    }
+}
+
+class Task
+{
+    /**
+     * @var string
+     */
+    private $pathToFile;
+
+    /**
+     * @var string
+     */
+    private $setName;
+
+    public function __construct(string $pathToFile, string $setName)
+    {
+        $this->pathToFile = $pathToFile;
+        $this->setName    = $setName;
+    }
+
+    /**
+     * @return string
+     */
+    public function getPathToFile(): string
+    {
+        return $this->pathToFile;
+    }
+
+    /**
+     * @return string
+     */
+    public function getSetName(): string
+    {
+        return $this->setName;
+    }
+}
+
+/**
+ * @see https://gist.github.com/divinity76/01ef9ca99c111565a72d3a8a6e42f7fb
+ *
+ * returns number of cpu cores
+ * Copyleft 2018, license: WTFPL
+ * @return int
+ */
+function num_cpu(): int
+{
+    if (defined('PHP_WINDOWS_VERSION_MAJOR')) {
+        $str = trim(shell_exec('wmic cpu get NumberOfCores 2>&1'));
+        if (!preg_match('/(\d+)/', $str, $matches)) {
+            throw new \RuntimeException ('wmic failed to get number of cpu cores on windows!');
+        }
+        return (( int ) $matches [1]);
+    }
+    $ret = @shell_exec('nproc');
+    if (is_string($ret)) {
+        $ret = trim($ret);
+        if (false !== ($tmp = filter_var($ret, FILTER_VALIDATE_INT))) {
+            return $tmp;
+        }
+    }
+    if (is_readable('/proc/cpuinfo')) {
+        $cpuinfo = file_get_contents('/proc/cpuinfo');
+        $count   = substr_count($cpuinfo, 'processor');
+        if ($count > 0) {
+            return $count;
+        }
+    }
+    throw new \LogicException('failed to detect number of CPUs!');
+}
```

Comment:
This looks quite a lot of classes. It might need testing and more solid checkers.

Could you add it to `utils/parallel-process-runner` as other more complex utils packages are?

Then just re-use the runner class in old `run_all_sets.php` file

Reply:
Sure

Reply:
Thank you